src/core/exec-invoke.c

```diff
@@ -3459,6 +3459,50 @@ static int close_remaining_fds(
         return close_all_fds(dont_close, n_dont_close);
 }
 
+static int close_and_invalidate_all_fds(
+                ExecParameters *params,
+                ExecRuntime *runtime,
+                const int *dont_close, size_t n_dont_close) {
+
+        int r;
+
+        assert(params);
+
+        r = close_all_fds(dont_close, n_dont_close);
+        if (r < 0)
+                return r;
+
+        /* We are done, invalidate all leftovers so that we don't double-close any of them in case there's a
+         * failure later and sd-executor cleanup-on-exit happens */
+
+        params->stdin_fd = params->stdout_fd = params->stderr_fd = params->user_lookup_fd = -EBADF;
+
+        if (runtime) {
+                runtime->ephemeral_storage_socket[0] = -EBADF;
+                runtime->ephemeral_storage_socket[1] = -EBADF;
+        }
+
+        if (runtime && runtime->shared) {
+                runtime->shared->netns_storage_socket[0] = -EBADF;
+                runtime->shared->netns_storage_socket[1] = -EBADF;
+                runtime->shared->ipcns_storage_socket[0] = -EBADF;
+                runtime->shared->ipcns_storage_socket[1] = -EBADF;
+        }
+
+        if (runtime && runtime->dynamic_creds) {
+                if (runtime->dynamic_creds->user) {
+                        runtime->dynamic_creds->user->storage_socket[0] = -EBADF;
+                        runtime->dynamic_creds->user->storage_socket[1] = -EBADF;
+                }
+                if (runtime->dynamic_creds->group) {
+                        runtime->dynamic_creds->group->storage_socket[0] = -EBADF;
+                        runtime->dynamic_creds->group->storage_socket[1] = -EBADF;
+                }
+        }
+
+        return r;
+}
+
```

Comment:
this is the wrong way round... please just close the runtime objects and others fds we apparently don't need anymore, first. And then close the left-over fds as always.

i.e. these things are not "left-overs" hence they shouldn't be closed like "left-overs". They are properly tracked by objects, and we should close those objects cleanly if we don't need them anymore instead of acting first, bringing things into an invalid state, and then trying to restore a valid state.

Reply:
but not only for logging - just check exec-invoke.c, both these objects are used plenty of times after the FDs are closed

Reply:
then add `exec_runtime_close()` which closes the fds but keeps the objects allocated.

and `exec_shared_runtime_close()` for the `runtime->shared` field. And `dynamic_creds_close()` for `runtime->dynamic_creds`. Both of which are called by `exec_runtime_close()`.

(unless of course you actually don't need one or both of these subject anymore, in which case you might as well close them entirely).

All I am saying is that I am not OK with blanket closing these fds, putting things into an invalid state, and then trying to restore a valid state after the kid already fell into the well. Close them correctly first please.

Reply:
well ok, but `close_all_fds()` has been there since forever... but doesn't matter, changed as suggested